### PR TITLE
Hnsw reclaim memory

### PR DESCRIPTION
**Pull request by alonre24**
Supplementary implementation for HNSW delete - reclaim memory by swapping the last element id with the deleted one. When index size is lower than index capacity in one block, we trim the data structures capacity.

### CODE REVIEW

File: src/VecSim/memory/vecsim_malloc.cpp [hunk 1]
```diff
@@ -36,12 +37,9 @@ void *VecSimAllocator::reallocate(void *p, size_t size) {
         return this->allocate(size);
     }
     size_t oldSize = getPointerAllocationSize(p);
-    if (oldSize >= size) {
-        return p;
-    }
     void *new_ptr = this->allocate(size);
     if (new_ptr) {
-        memcpy(new_ptr, p, oldSize);
+        memcpy(new_ptr, p, MIN(oldSize, size));
```
Comment (DvirDukhan) [inline src/VecSim/memory/vecsim_malloc.cpp:42]: why the change here?

File: src/VecSim/algorithms/hnsw/hnswlib.h [hunk 2]
```diff
@@ -66,6 +66,10 @@ class HierarchicalNSW : public VecsimBaseObject {
 
     // Index state
     size_t cur_element_count;
+    // TODO: after introducing the memory reclaim upon delete, max_id is redundant since the valid
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnswlib.h:69]: Why not? `max_id` will always be now `cur_element_count-1`

File: src/VecSim/memory/vecsim_malloc.cpp [hunk 3]
```diff
@@ -2,6 +2,7 @@
 #include <stdlib.h>
 #include <memory>
 #include <string.h>
+#include <sys/param.h>
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/memory/vecsim_malloc.cpp:5]: For using the MIN macro

File: tests/unit/test_hnswlib.cpp [hunk 4]
```diff
@@ -1424,24 +1424,40 @@ TEST_F(HNSWLibTest, testSizeEstimation) {
...
 
+    float vec[dim];
     for (size_t i = 0; i < n; i++) {
+        for (size_t j = 0; j < dim; j++) {
+            vec[j] = (float)i;
+        }
         VecSimIndex_AddVector(index, vec, i);
     }
+
+    // Estimate the memory delta of adding a fool new block.
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnswlib.cpp:1449]: yes... :sweat_smile:

File: src/VecSim/memory/vecsim_malloc.cpp (hunk 1 above)
Comment (alonre24) [reply src/VecSim/memory/vecsim_malloc.cpp:42]: because if `oldSize>size`, we don't want to copy `oldSize` bytes to the new buffer of size `size` (we'll get segmentation fault...)

File: src/VecSim/algorithms/hnsw/hnswlib.h [hunk 5]
```diff
@@ -794,10 +917,9 @@ HierarchicalNSW<dist_t>::~HierarchicalNSW() {
  */
 template <typename dist_t>
 void HierarchicalNSW<dist_t>::resizeIndex(size_t new_max_elements) {
-    if (new_max_elements < cur_element_count)
-        throw std::runtime_error(
-            "Cannot resize, max element is less than the current number of elements");
     element_levels_.resize(new_max_elements);
+    element_levels_.shrink_to_fit();
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnswlib.h:921]: no... (I tested it)

File: src/VecSim/algorithms/hnsw/hnswlib.h [hunk 6]
```diff
@@ -795,14 +795,9 @@ void HierarchicalNSW<dist_t>::SwapLastIdWithDeletedId(tableint element_internal_
...
-    if (element_levels_[last_element_internal_id] > 0) {
-        linkLists_[element_internal_id] = (char *)this->allocator->allocate(
-            size_links_per_element_ * element_levels_[last_element_internal_id] + 1);
-        memcpy(linkLists_[element_internal_id], linkLists_[last_element_internal_id],
-               size_links_per_element_ * element_levels_[last_element_internal_id] + 1);
-        this->allocator->free_allocation(linkLists_[last_element_internal_id]);
-    }
+    // swap pointer of higher levels links
+    linkLists_[element_internal_id] = linkLists_[last_element_internal_id];
+    linkLists_[last_element_internal_id] = nullptr;
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnswlib.h:800]: leak? you don't need to free it?

**Review by DvirDukhan — APPROVED** [review]
